File: eyot-instance-host/src/protocol.ts
/** Tunnel envelope shared with eyot-backend tunnel protocol. */

/**
 * Versioned message types, aligned 1:1 with the backend Tunnel protocol
 * (`eyot-backend/app/services/tunnel/protocol.py`). Dotted names are the
 * wire contract; new message types must follow the same `area.name` shape.
 */
export enum TunnelMessageType {
  // Backend -> Host
  AUTH_OK = "auth.ok",
  AUTH_ERROR = "auth.error",
  CHAT_REQUEST = "chat.request",
  CONTROL = "control",
  PING = "ping",

  // Host -> Backend
  AUTH = "auth",
  CHAT_RESPONSE_CHUNK = "chat.response.chunk",
  CHAT_RESPONSE_DONE = "chat.response.done",
  CHAT_RESPONSE_ERROR = "chat.response.error",
  CHAT_RESPONSE_ACTIVITY = "chat.response.activity",
  PONG = "pong",
}

export interface TunnelMessage {
  id: string;
  type: string;
  reply_to?: string;
  turn_id?: string;
  payload: Record<string, unknown>;
  ts: number;
}

export function newId(): string {
  return crypto.randomUUID();
}

export function makeMessage(
  type: string,
  payload: Record<string, unknown> = {},
  extra: Partial<Pick<TunnelMessage, "reply_to" | "turn_id">> = {},
): TunnelMessage {
  return {
    id: newId(),
    type,
    payload,
    ts: Date.now(),
    ...extra,
  };
}

export function httpToWs(url: string): string {
  return url
    .replace(/^https:\/\//i, "wss://")
    .replace(/^http:\/\//i, "ws://")
    .replace(/\/+$/, "");
}

export function deriveTunnelUrl(apiUrl: string, override?: string): string {
  if (override && override.trim()) {
    return override.trim().replace(/\/+$/, "");
  }
  return `${httpToWs(apiUrl)}/api/v1/tunnel/connect`;
}

File: cocoa-backend/app/services/tunnel/protocol.py
"""Eyot Tunnel WebSocket message protocol (PRD-v3.5 / P14b revival).

Frame shape aligns with Composer SSE (`chat.response.*`) and nodeskclaw Tunnel.
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class TunnelMessageType(str, Enum):
    # Backend -> Instance Host
    AUTH_OK = "auth.ok"
    AUTH_ERROR = "auth.error"
    CHAT_REQUEST = "chat.request"
    CONTROL = "control"
    PING = "ping"

    # Instance Host -> Backend
    AUTH = "auth"
    CHAT_RESPONSE_CHUNK = "chat.response.chunk"
    CHAT_RESPONSE_DONE = "chat.response.done"
    CHAT_RESPONSE_ERROR = "chat.response.error"
    CHAT_RESPONSE_ACTIVITY = "chat.response.activity"
    PONG = "pong"


@dataclass
class TunnelMessage:
    type: str
    payload: dict[str, Any] = field(default_factory=dict)
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    reply_to: str | None = None
    turn_id: str | None = None
    ts: int = field(default_factory=lambda: int(time.time() * 1000))

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "id": self.id,
            "type": self.type,
            "payload": self.payload,
            "ts": self.ts,
        }
        if self.reply_to:
            d["reply_to"] = self.reply_to
        if self.turn_id:
            d["turn_id"] = self.turn_id
        return d

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> TunnelMessage:
        return cls(
            id=str(raw.get("id") or uuid.uuid4()),
            type=str(raw.get("type") or ""),
            payload=dict(raw.get("payload") or {}),
            reply_to=raw.get("reply_to") or raw.get("replyTo"),
            turn_id=raw.get("turn_id") or raw.get("turnId"),
            ts=int(raw.get("ts") or time.time() * 1000),
        )
